pkgs/by-name/ni/nixos-render-docs-redirects/src/nixos_render_docs_redirects/__init__.py

```diff
@@ -0,0 +1,86 @@
+import json
+import argparse
+from pathlib import Path
+
+
+def add_new_identifier(redirects: dict[str, list[str]], identifier: str, path: str) -> dict[str, list[str]]:
+    if anchor in redirects:
+        print("The identifier in the given element already exists.")
+        exit(1)
+    redirects[anchor] = [args.element]
+    return redirects
+
+
+def redirect_to_new_path(
+    redirects: dict[str, list[str]],
+    old_path: str,
+    old_anchor: str,
+    new_path: str,
+    new_anchor: str
+) -> dict[str, list[str]]:
+    if new_anchor != old_anchor:
+        new_redirects = {}
+        for key, value in redirects.items():
+            if key == old_anchor:
+                new_redirects[new_anchor] = value
+                continue
+            new_redirects[key] = value
+        redirects = new_redirects
+    redirects[new_anchor].insert(0, f"{new_path}#{new_anchor}")
+    return redirects
+
+
+def main():
+    parser = argparse.ArgumentParser(description="redirects manipulation for nixos manuals")
+    parser.add_argument("-f", "--file", type=Path)
+    commands = parser.add_subparsers(dest="command", required=True)
```

Comment:
Unless it's needed for testing, I'd wish the user wouldn't have to bother with specifying the redirects file. It could be hard-coded in the repo's development shell for instance, such that in the shell all you'd ever have to do would be something like

```bash
add-content foo index.html#foo
move-content bar hello.html#bar 
rename-identifier x y
```

Reply:
Currently (actually only internally relevant) identifiers and anchors are the same, which is a trade-off for syntactic simplicity. We *could* leverage that to simplify even further and make it something like this:

An identifier `foo` was added, and will render in output file `index.html`:
 
```bash
add-content foo index.html
```

An identifier `bar` rendered somewhere else before, but will now move to `hello.html`:

```bash
move-content bar hello.html
```

An identifier `x` was rendering to `<something>#x` but is now renamed to `y` and will thus render to `<something>#y`, which requires a client-side redirect.

```bash
rename-identifier x y
```

There's a composition of `move-content` and `rename-identifier`, which we may or may not want as a unit for convenience.

Another thing we've already seen happening is removing an identifier and redirecting its old locations to a new identifier:

```
remove-and-redirect p q
```

This would take all locations of `p` and prepend them to `q`'s locations, and remove the entry for `p`.

Reply:
> Unless it's needed for testing, I'd wish the user wouldn't have to bother with specifying the redirects file. It could be hard-coded in the repo's development shell for instance

We have 2 redirects files, which means a single hard coded value won't work. I think there are 3 options (from what I can gather):
- Accept a file path (current approach)
- Accept a flag nixos/nixpkgs and it uses the hardcoded path based on this.
- Read the current working directory of the user, and if it's `doc/` or `nixos/doc/manual`, we have our answer and just fall back to either of the 2 options above otherwise. (I don't think this is a good approach)

Reply:
As for the commands, I like the idea of keeping their signatures in alignment of how a user would do those operations in the docs.

> There's a composition of `move-content` and `rename-identifier`, which we may or may not want as a unit for convenience.

Which seems to be the 4th command you sent, but it doesn't seem like that's the one you meant from your text; could you clarify?

Reply:
One composition I can see is with `move-content` having a flag `--identifier-name` which would also rename the identifier. WDYT?

Reply:
> We have 2 redirects files, which means a single hard coded value won't work

Why can't we have executables in the shell for each manual? Don't remember but AFAIK the NixOS manual doesn't even have a devshell currently, and in any case since it lives in a different directory, it may as well have a separate one, where the helper commands are wrapped with the right parameter.

> Which seems to be the 4th command you sent, but it doesn't seem like that's the one you meant from your text; could you clarify?

No, or at most maybe, because removing is a separate thing. I meant something like what I've seen in one of the recent docs PRs where you rename an identifier and change the output location in one commit.


> One composition I can see is with move-content having a flag --identifier-name which would also rename the identifier.

Yes, something like that. Looks a bit cumbersome syntactically, but the semantics is right. It *can* be two helper operations, but if it's a common thing, having one helper operation would be convenient. And since we're generating the suggestions anyway, the only requirement is that the commands have little visual noise while roughly indicating what they're going to do. Otherwise they can be almost arbitrarily obscure.